src/MaIN.Core/Hub/Contexts/ModelContext.cs

```diff
@@ -0,0 +1,299 @@
+
+        try
+        {
+            using var response = await httpClient.GetAsync(url, HttpCompletionOption.ResponseHeadersRead, cancellationToken);
+            response.EnsureSuccessStatusCode();
+
+            await DownloadWithProgressAsync(response, filePath, fileName, cancellationToken);
+        }
+        catch (Exception ex)
+        {
+            Console.WriteLine($"Download failed: {ex.Message}");
+            
+            if (!File.Exists(filePath)) throw;
+            File.Delete(filePath);
+            throw;
+        }
+    }
+
+    private async Task DownloadWithProgressAsync(HttpResponseMessage response, string filePath, string fileName, CancellationToken cancellationToken)
+    {
+        var totalBytes = response.Content.Headers.ContentLength;
+        var totalBytesRead = 0L;
+        var buffer = new byte[DefaultBufferSize];
+        var lastProgressUpdate = DateTime.Now;
+        var startTime = DateTime.Now;
+
+        if (totalBytes.HasValue)
+        {
+            Console.WriteLine($"File size: {FormatBytes(totalBytes.Value)}");
+        }
+
+        await using var fileStream = new FileStream(filePath, FileMode.Create, FileAccess.Write, FileShare.None, FileStreamBufferSize);
+        await using var contentStream = await response.Content.ReadAsStreamAsync(cancellationToken);
+
+        while (true)
+        {
+            var bytesRead = await contentStream.ReadAsync(buffer, 0, buffer.Length, cancellationToken);
+            if (bytesRead == 0) break;
+
+            await fileStream.WriteAsync(buffer, 0, bytesRead, cancellationToken);
+            totalBytesRead += bytesRead;
+
+            if (ShouldUpdateProgress(lastProgressUpdate))
+            {
+                ShowProgress(totalBytesRead, totalBytes, startTime);
+                lastProgressUpdate = DateTime.Now;
+            }
+        }
+
+        ShowFinalProgress(totalBytesRead, totalBytes, startTime, fileName);
+    }
+
+    private void DownloadModelSync(string url, string fileName)
+    {
+        var filePath = GetModelFilePath(fileName);
+        
+        Console.WriteLine($"Starting download of {fileName}...");
+
+        using var webClient = CreateConfiguredWebClient();
+        var startTime = DateTime.Now;
+        var lastProgressUpdate = DateTime.Now;
+        
+        webClient.DownloadProgressChanged += (sender, e) =>
+        {
+            if (ShouldUpdateProgress(lastProgressUpdate))
+            {
+                ShowProgress(e.BytesReceived, e.TotalBytesToReceive > 0 ? e.TotalBytesToReceive : null, startTime);
+                lastProgressUpdate = DateTime.Now;
+            }
+        };
+        
+        webClient.DownloadFileCompleted += (sender, e) =>
+        {
+            if (e.Error != null)
+            {
+                Console.WriteLine($"\nDownload failed: {e.Error.Message}");
+            }
+            else
+            {
+                var totalTime = DateTime.Now - startTime;
```

Comment:
Use a Stopwatch instead of DateTime operations. It's more lightweight and precise for such tasks.